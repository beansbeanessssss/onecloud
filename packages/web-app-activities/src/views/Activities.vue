<template>
  <main class="oc-height-1-1 app-content oc-width-1-1">
    <app-loading-spinner v-if="loadActivitiesTask.isRunning" />
    <timeline
      v-else-if="loadActivitiesTask.lastSuccessful.value.length"
      :activities="loadActivitiesTask.lastSuccessful.value"
    ></timeline>
    <no-content-message v-else icon="folder" class="activities-empty">
      <template #message>
        <p class="oc-text-muted">
          <span v-translate>No activities</span>
        </p>
      </template>
    </no-content-message>
  </main>
</template>

<script lang="ts">
import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import Timeline from '../components/Timeline.vue'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import { defineComponent } from '@vue/runtime-core'
import { useStore, useRequest } from 'web-pkg/src/composables'
import { useTranslations } from 'web-pkg/src/composables/translations'
import { useTask } from 'vue-concurrency'
import { DateTime } from 'luxon'
import { formatRelativeDateFromDateTime } from 'web-pkg/src/helpers'

const buildActivity = ({ type, dateTime, data }, { $gettext, $gettextInterpolate }) => {
  const activity = {
    title: data.filename,
    icon: {
      name: 'check',
      fill: 'line',
      variation: 'passive'
    },
    body: '',
    tags: [formatRelativeDateFromDateTime(DateTime.fromISO(dateTime), navigator.language)]
  }

  switch (type) {
    case 'VirusscanFinished':
      activity.tags.push('virus-scan')

      if (data.infected) {
        activity.icon.name = 'file-warning'
        activity.icon.variation = 'warning'
      }

      if (data.description) {
        activity.body = $gettextInterpolate(
          $gettext(
            'Virus "%{description}" detected. Please contact your administrator for more information.'
          ),
          { description: data.description },
          true
        )
      }

      if (data.outcome === 'delete') {
        activity.tags.push('deleted')
      }
  }

  return activity
}

export default defineComponent({
  components: {
    AppLoadingSpinner,
    Timeline,
    NoContentMessage
  },
  setup() {
    const store = useStore()
    const { makeRequest } = useRequest()
    const translationHelpers = useTranslations()
    const loadActivitiesTask = useTask(function* (signal) {
      const { data } = yield makeRequest(
        'GET',
        `${store.getters.configuration.server}experimental/activities`,
        {}
      )

      return data.reduce((acc, act) => [...acc, buildActivity(act, translationHelpers)], [])
    })

    loadActivitiesTask.perform()

    return {
      loadActivitiesTask
    }
  }
})
</script>

<style lang="scss" scoped>
main {
  overflow-y: auto;
}
</style>
