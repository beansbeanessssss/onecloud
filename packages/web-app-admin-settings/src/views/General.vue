<template>
  <div>
    <app-template ref="template" :breadcrumbs="breadcrumbs">
      <template #mainContent>
        <div class="oc-px-m">
          <Info />
        </div>
      </template>
    </app-template>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue'
import AppTemplate from '../components/AppTemplate.vue'
import Info from '../components/General/Info.vue'

export default defineComponent({
  components: {
    Info,
    AppTemplate
  },
  setup() {
    const template = ref()

    return {
      template
    }
  },
  computed: {
    breadcrumbs() {
      return [
        { text: this.$gettext('Administration Settings'), to: { path: '/admin-settings' } },
        {
          text: this.$gettext('General')
        }
      ]
    }
  }
})
</script>
