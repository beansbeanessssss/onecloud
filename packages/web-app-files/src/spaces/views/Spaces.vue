<template>
  <div class="oc-p-s">
    <h2 v-text="$gettext('Spaces')" />
    <div>
      <span v-text="$gettext('Access all project related files in one place.')" />
      <a href="#" v-text="$gettext('Learn more about spaces.')" />
      <h3 v-text="$gettext('Your spaces')" />
      <hr class="oc-mb-s" />
    </div>
    <list-loader v-if="loadSpacesTask.isRunning" />
    <template v-else>
      <no-content-message
        v-if="!spaces.length"
        id="files-spaces-empty"
        class="files-empty"
        icon="layout-grid"
      >
        <template #message>
          <span v-translate>You don't have access to any spaces</span>
        </template>
      </no-content-message>
      <div v-else class="spaces-list">
        <div
          class="
            uk-grid-match uk-grid-column-small uk-grid-row-large uk-text-center uk-child-width-1-3@s
          "
          uk-grid
        >
          <a v-for="space in spaces" :key="space.id" href="#">
            <span class="spaces-list-card oc-border uk-card uk-card-default">
              <span class="uk-card-media-top oc-border-b">
                <img
                  :src="space.image || defaultImg"
                  :class="{ 'oc-px-m': !space.image, 'oc-py-m': !space.image }"
                  alt=""
                />
              </span>
              <span class="uk-card-body">
                <h3 class="uk-card-title">{{ space.name }}</h3>
              </span>
            </span>
          </a>
        </div>
      </div>
    </template>
  </div>
</template>

<script>
import NoContentMessage from '../../components/FilesList/NoContentMessage.vue'
import ListLoader from '../../components/FilesList/ListLoader.vue'
import { ref } from '@vue/composition-api'
import { useStore } from '../../composables'
import { spacesSDK } from '../sdk'
import { useTask } from 'vue-concurrency'

export default {
  components: {
    NoContentMessage,
    ListLoader
  },
  setup() {
    const store = useStore()
    const spaces = ref([])
    const defaultImg = store.getters.configuration.theme.spaces?.defaultImg

    const loadSpacesTask = useTask(function* (signal) {
      const localSDK = spacesSDK(store.getters.configuration.server, store.getters.getToken)
      const myDrives = (yield localSDK.drives.listMyDrives()).data?.value

      if (!myDrives) {
        return
      }

      spaces.value = myDrives.filter((drive) => drive.driveType === 'project')
    })

    loadSpacesTask.perform()

    return {
      spaces,
      defaultImg,
      loadSpacesTask
    }
  }
}
</script>

<style lang="scss">
#files-spaces-empty {
  height: 50vh;
}

.spaces-list {
  &-card {
    box-shadow: none !important;
  }

  .uk-card-media-top {
    display: inline-block;
    width: 100%;
    background-color: var(--oc-color-background-muted);
  }

  .uk-card-media-top img {
    max-height: 150px;
  }
}
</style>
