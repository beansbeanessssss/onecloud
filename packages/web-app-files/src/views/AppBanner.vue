<template>
  <div class="app-banner hide-desktop">
    <a href="javascript:void(0);" class="app-banner-exit" aria-label="Close"></a>
    <div
      class="app-banner-icon"
      style="
        background-image: url('https://is1-ssl.mzstatic.com/image/thumb/Purple116/v4/87/4c/3c/874c3c90-d58d-837f-63fe-6a7dae6fd320/AppIcon-1x_U007emarketing-0-7-0-sRGB-85-220.png/460x0w.png');
      "
    ></div>
    <div class="info-container">
      <div>
        <div class="app-title">ownCloud</div>
        <div class="app-publisher">ownCloud GmbH</div>
        <div class="app-additional-info">FREE - On the App Store</div>
      </div>
    </div>
    <a
      :href="generateAppUrl(fileId)"
      target="_blank"
      class="app-banner-cta"
      rel="noopener"
      aria-label="VIEW"
      >VIEW</a
    >
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref, unref } from 'vue'
import { useStore } from 'web-pkg'

export default defineComponent({
  components: {},
  props: {
    fileId: {
      type: String,
      required: true
    }
  },
  setup() {
    const store = useStore()
    const mobileAppScheme = store.getters.configuration.commonTheme.mobileAppScheme || 'owncloud'

    const generateAppUrl = (fileId: string) => {
      return `${mobileAppScheme}://${window.location.host}/f/${fileId}`
    }
    return {
      generateAppUrl
    }
  }
})
</script>

<style scoped lang="scss">
.hide-desktop {
  @media (min-width: 768px) {
    display: none;
  }
}

.app-banner {
  position: absolute;
  top: 0;
  left: 0;
  overflow-x: hidden;
  width: 100%;
  height: 84px;
  background: #f3f3f3;
  font-family: Helvetica, sans, sans-serif;
}

.info-container {
  position: absolute;
  top: 10px;
  left: 104px;
  display: flex;
  overflow-y: hidden;
  width: 60%;
  height: 64px;
  align-items: center;
  color: #000;
}

.app-banner-icon {
  position: absolute;
  top: 10px;
  left: 30px;
  width: 64px;
  height: 64px;
  border-radius: 15px;
  background-size: 64px 64px;
}

.app-banner-cta {
  position: absolute;
  top: 32px;
  right: 10px;
  z-index: 1;
  display: block;
  padding: 0 10px;
  min-width: 10%;
  border-radius: 5px;
  background: #f3f3f3;
  color: #1474fc;
  font-size: 18px;
  text-align: center;
  text-decoration: none;
}

.app-title {
  font-size: 14px;
}

.app-publisher,
.app-additional-info {
  font-size: 12px;
}
.app-banner-exit::after {
  transform: rotate(-45deg);
}

.app-banner-exit::before,
.app-banner-exit::after {
  position: absolute;
  width: 1px;
  height: 12px;
  background: #767676;
  content: ' ';
}

.app-banner-exit::before {
  transform: rotate(45deg);
}

.app-banner-exit {
  position: absolute;
  top: calc(50% - 6px);
  left: 9px;
  display: block;
  margin: 0;
  width: 12px;
  height: 12px;
  border: 0;
  text-align: center;
}
</style>
