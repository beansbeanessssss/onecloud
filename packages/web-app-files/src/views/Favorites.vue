<template>
  <div class="oc-flex">
    <files-view-wrapper>
      <app-bar :view-modes="viewModes" :is-side-bar-open="isSideBarOpen" />
      <app-loading-spinner v-if="areResourcesLoading" />
      <template v-else>
        <no-content-message
          v-if="isEmpty"
          id="files-favorites-empty"
          class="files-empty"
          icon="star"
        >
          <template #message>
            <span v-translate>There are no resources marked as favorite</span>
          </template>
        </no-content-message>
        <resource-table
          v-else
          id="files-favorites-table"
          v-model:selectedIds="selectedResourcesIds"
          resource-type="file"
          :are-paths-displayed="true"
          :are-thumbnails-displayed="displayThumbnails"
          :resources="paginatedResources"
          :header-position="fileListHeaderY"
          :sort-by="sortBy"
          :sort-dir="sortDir"
          @file-click="triggerDefaultAction"
          @row-mounted="rowMounted"
          @sort="handleSort"
        >
          <template #quickActions="props">
            <quick-actions class="oc-visible@s" :item="props.resource" />
          </template>
          <template #contextMenu="{ resource }">
            <context-actions
              v-if="isResourceInSelection(resource)"
              :action-options="{ space: getMatchingSpace(resource), resources: selectedResources }"
            />
          </template>
          <template #footer>
            <pagination :pages="paginationPages" :current-page="paginationPage" />
            <list-info
              v-if="paginatedResources.length > 0"
              class="oc-width-1-1 oc-my-s"
              :files="totalResourcesCount.files"
              :folders="totalResourcesCount.folders"
              :size="totalResourcesSize"
            />
          </template>
        </resource-table>
      </template>
    </files-view-wrapper>
    <file-side-bar
      :is-open="isSideBarOpen"
      :active-panel="sideBarActivePanel"
      :space="selectedResourceSpace"
    />
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, onBeforeUnmount, onMounted, ref } from 'vue'
import { debounce } from 'lodash-es'

import { Resource } from '@ownclouders/web-client'
import { VisibilityObserver, useConfigStore, useResourcesStore } from '@ownclouders/web-pkg'
import { ImageDimension } from '@ownclouders/web-pkg'
import { AppLoadingSpinner } from '@ownclouders/web-pkg'
import { FileSideBar, NoContentMessage } from '@ownclouders/web-pkg'
import { Pagination } from '@ownclouders/web-pkg'
import { eventBus } from '@ownclouders/web-pkg'
import { useGetMatchingSpace, ViewModeConstants } from '@ownclouders/web-pkg'

import { AppBar } from '@ownclouders/web-pkg'
import QuickActions from '../components/FilesList/QuickActions.vue'
import ListInfo from '../components/FilesList/ListInfo.vue'
import { ContextActions } from '@ownclouders/web-pkg'
import { ResourceTable } from '@ownclouders/web-pkg'
import FilesViewWrapper from '../components/FilesViewWrapper.vue'
import { useResourcesViewDefaults } from '../composables'
import { useFileActions } from '@ownclouders/web-pkg'
import { storeToRefs } from 'pinia'

const visibilityObserver = new VisibilityObserver()

export default defineComponent({
  components: {
    FilesViewWrapper,
    AppBar,
    FileSideBar,
    ResourceTable,
    QuickActions,
    AppLoadingSpinner,
    Pagination,
    NoContentMessage,
    ListInfo,
    ContextActions
  },

  setup() {
    const { getMatchingSpace } = useGetMatchingSpace()
    const configStore = useConfigStore()
    const { options: configOptions } = storeToRefs(configStore)

    const resourcesStore = useResourcesStore()
    const { updateResourceField } = resourcesStore
    const { totalResourcesCount, totalResourcesSize } = storeToRefs(resourcesStore)

    const viewModes = computed(() => [
      ViewModeConstants.condensedTable,
      ViewModeConstants.default,
      ViewModeConstants.tilesView
    ])

    const loadResourcesEventToken = ref(null)

    onMounted(() => {
      loadResourcesEventToken.value = eventBus.subscribe(
        'app.files.list.removeFromFavorites',
        (resourceId: string) => {
          resourcesStore.removeResources([{ id: resourceId }] as Resource[])
        }
      )
    })

    onBeforeUnmount(() => {
      visibilityObserver.disconnect()
      eventBus.unsubscribe('app.files.list.removeFromFavorites', loadResourcesEventToken)
    })

    return {
      ...useFileActions(),
      ...useResourcesViewDefaults<Resource, any, any[]>(),
      configOptions,
      getMatchingSpace,
      viewModes,
      updateResourceField,
      totalResourcesCount,
      totalResourcesSize
    }
  },

  computed: {
    isEmpty() {
      return this.paginatedResources.length < 1
    },

    displayThumbnails() {
      return !this.configOptions.disablePreviews
    }
  },

  async created() {
    await this.loadResourcesTask.perform()
    this.scrollToResourceFromRoute(this.paginatedResources, 'files-app-bar')
  },

  methods: {
    rowMounted(resource: Resource, component) {
      if (!this.displayThumbnails) {
        return
      }

      const loadPreview = async () => {
        const preview = await this.$previewService.loadPreview(
          {
            space: this.getMatchingSpace(resource),
            resource,
            dimensions: ImageDimension.Thumbnail
          },
          true
        )
        if (preview) {
          this.updateResourceField({ id: resource.id, field: 'thumbnail', value: preview })
        }
      }

      const debounced = debounce(({ unobserve }) => {
        unobserve()
        loadPreview()
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    }
  }
})
</script>
