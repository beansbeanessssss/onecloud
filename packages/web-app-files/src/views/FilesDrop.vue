<template>
  <div id="files-drop-container" class="oc-height-1-1 oc-flex oc-flex-column oc-flex-between">
    <div v-if="dragareaEnabled" class="dragarea" />
    <h1 class="oc-invisible-sr">{{ pageTitle }}</h1>
    <div class="oc-p oc-height-1-1">
      <div v-if="loading" key="loading-drop" class="oc-flex oc-flex-column oc-flex-middle">
        <h2 class="oc-login-card-title">
          <span v-text="$gettext('Loading public linkâ€¦')" />
        </h2>
        <oc-spinner :aria-hidden="true" />
      </div>
      <div v-else key="loaded-drop" class="oc-flex oc-flex-column oc-flex-middle oc-height-1-1">
        <div class="oc-text-center oc-width-1-1 oc-width-xxlarge@m">
          <h2 v-text="title" />
          <resource-upload
            id="files-drop-zone"
            ref="fileUpload"
            class="oc-flex oc-flex-middle oc-flex-center oc-placeholder"
            :btn-label="$gettext('Drop files here to upload or click to select file')"
          />
          <div id="previews" hidden />
        </div>
        <div v-if="errorMessage" class="oc-text-center">
          <h2>
            <span v-text="$gettext('An error occurred while loading the public link')" />
          </h2>
          <p class="oc-m-rm" v-text="errorMessage" />
        </div>
      </div>
    </div>
    <div class="oc-text-center">
      <p v-text="configuration.currentTheme.general.slogan" />
    </div>
  </div>
</template>

<script lang="ts">
import { mapGetters } from 'vuex'
import { DavProperties, DavProperty } from 'web-client/src/webdav/constants'
import { createLocationPublic } from '../router'

import ResourceUpload from '../components/AppBar/Upload/ResourceUpload.vue'
import { defineComponent, getCurrentInstance, onMounted, onBeforeUnmount, ref, unref } from 'vue'
import { useUpload } from 'web-runtime/src/composables/upload'
import * as uuid from 'uuid'
import {
  useClientService,
  usePublicLinkPassword,
  usePublicLinkToken,
  useStore
} from 'web-pkg/src/composables'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { linkRoleUploaderFolder } from 'web-client/src/helpers/share'
import { useService } from 'web-pkg/src/composables/service'
import { UppyService } from 'web-runtime/src/services/uppyService'
import { useAuthService } from 'web-pkg/src/composables/authContext/useAuthService'
import { useRouter } from 'vue-router'

export default defineComponent({
  components: {
    ResourceUpload
  },
  setup() {
    const instance = getCurrentInstance().proxy as any
    const uppyService = useService<UppyService>('$uppyService')
    const store = useStore()
    const router = useRouter()
    const authService = useAuthService()
    const { owncloudSdk } = useClientService()
    const publicToken = usePublicLinkToken({ store })
    const publicLinkPassword = usePublicLinkPassword({ store })

    const filesSelectedSub = ref()
    const dragOver = ref()
    const dragOut = ref()
    const drop = ref()
    const dragareaEnabled = ref(false)
    const loading = ref(true)
    const errorMessage = ref(null)

    const hideDropzone = () => {
      dragareaEnabled.value = false
    }
    const onDragOver = (event) => {
      dragareaEnabled.value = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }

    const resolvePublicLink = () => {
      loading.value = true
      owncloudSdk.publicFiles
        .list(unref(publicToken), unref(publicLinkPassword), DavProperties.PublicLink, '0')
        .then((files) => {
          // Redirect to files list if the link doesn't have role "uploader"
          const sharePermissions = parseInt(files[0].getProperty(DavProperty.PublicLinkPermission))
          if (linkRoleUploaderFolder.bitmask(false) !== sharePermissions) {
            router.replace(
              createLocationPublic('files-public-link', {
                params: { driveAliasAndItem: `public/${unref(publicToken)}` }
              })
            )
            return
          }
          // TODO: There is not this.share?!
          // this.share = files[0]
        })
        .catch((error) => {
          // likely missing password, redirect to public link password prompt
          if (error.statusCode === 401) {
            return authService.handleAuthError(unref(router.currentRoute))
          }
          console.error(error)
          errorMessage.value = error
        })
        .finally(() => {
          loading.value = false
        })
    }

    onMounted(() => {
      dragOver.value = eventBus.subscribe('drag-over', onDragOver)
      dragOut.value = eventBus.subscribe('drag-out', hideDropzone)
      drop.value = eventBus.subscribe('drop', hideDropzone)
      filesSelectedSub.value = uppyService.subscribe('filesSelected', instance.onFilesSelected) // FIXME
      uppyService.useDropTarget({
        targetSelector: '#files-drop-container',
        uppyService
      })
      resolvePublicLink()
    })

    onBeforeUnmount(() => {
      eventBus.unsubscribe('drag-over', unref(dragOver))
      eventBus.unsubscribe('drag-out', unref(dragOut))
      eventBus.unsubscribe('drop', unref(drop))
      uppyService.unsubscribe('filesSelected', unref(filesSelectedSub))
      uppyService.removeDropTarget()
    })

    return {
      ...useUpload({
        uppyService
      }),
      dragareaEnabled,
      loading,
      errorMessage
    }
  },
  computed: {
    ...mapGetters(['configuration']),
    pageTitle() {
      return this.$gettext(this.$route.meta.title)
    },
    publicLinkToken() {
      return this.$route.params.token
    },
    title() {
      // share might not be loaded
      if (this.share) {
        const translated = this.$gettext('%{owner} shared this folder with you for uploading')
        return this.$gettextInterpolate(
          translated,
          { owner: this.share.getProperty(this.$client.publicFiles.PUBLIC_LINK_SHARE_OWNER) },
          true
        )
      }
      return ''
    },
    url() {
      return this.$client.publicFiles.getFileUrl(this.publicLinkToken) + '/'
    }
  },
  methods: {
    onFilesSelected(files) {
      this.$uppyService.publish('uploadStarted')

      const uppyResources = files.map((file) => ({
        source: 'FileDrop',
        name: file.name,
        type: file.type,
        data: file,
        meta: {
          tusEndpoint: this.url,
          relativePath: file.webkitRelativePath || file.relativePath || '',
          uploadId: uuid.v4()
        }
      }))

      this.$uppyService.publish('addedForUpload', uppyResources)
      this.$uppyService.uploadFiles(uppyResources)
    }
  }
})
</script>

<style lang="scss">
#files-drop-container {
  position: relative;
  background: transparent;
  border: 1px dashed var(--oc-color-input-border);
  margin: var(--oc-space-xlarge);
}
.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}
</style>
