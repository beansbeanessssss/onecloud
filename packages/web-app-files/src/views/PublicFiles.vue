<template>
  <div class="oc-flex">
    <keyboard-actions :paginated-resources="paginatedResources" />
    <files-view-wrapper>
      <app-bar
        :has-bulk-actions="true"
        :breadcrumbs="breadcrumbs"
        :breadcrumbs-context-actions-items="[currentFolder]"
        :show-actions-on-selection="true"
        :side-bar-open="sideBarOpen"
      >
        <template #actions="{ limitedScreenSpace }">
          <create-and-upload :limited-screen-space="limitedScreenSpace" />
        </template>
      </app-bar>
      <app-loading-spinner v-if="areResourcesLoading" />
      <template v-else>
        <not-found-message v-if="folderNotFound" class="files-not-found oc-height-1-1" />
        <no-content-message
          v-else-if="isEmpty"
          id="files-public-list-empty"
          class="files-empty"
          icon="folder"
        >
          <template #message>
            <span v-translate>There are no resources in this folder</span>
          </template>
          <template v-if="currentFolder.canCreate()" #callToAction>
            <span v-translate data-testid="public-files-call-to-action">
              Drag files and folders here or use the "New" or "Upload" buttons to add files
            </span>
          </template>
        </no-content-message>
        <resource-table
          v-else
          id="files-public-files-table"
          v-model="selectedResourcesIds"
          class="files-table"
          :class="{ 'files-table-squashed': sideBarOpen }"
          :fields-displayed="['name', 'size', 'mdate']"
          :are-thumbnails-displayed="displayThumbnails"
          :resources="paginatedResources"
          :target-route="targetRoute"
          :header-position="fileListHeaderY"
          :sort-by="sortBy"
          :sort-dir="sortDir"
          :drag-drop="true"
          @fileDropped="fileDropped"
          @fileClick="$_fileActions_triggerDefaultAction"
          @rowMounted="rowMounted"
          @sort="handleSort"
        >
          <template #contextMenu="{ resource }">
            <context-actions v-if="isResourceInSelection(resource)" :items="selectedResources" />
          </template>
          <template #footer>
            <pagination :pages="paginationPages" :current-page="paginationPage" />
            <list-info
              v-if="paginatedResources.length > 0"
              class="oc-width-1-1 oc-my-s"
              :files="totalFilesCount.files"
              :folders="totalFilesCount.folders"
              :size="totalFilesSize"
            />
          </template>
        </resource-table>
      </template>
    </files-view-wrapper>
    <side-bar :open="sideBarOpen" :active-panel="sideBarActivePanel" />
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions, mapMutations } from 'vuex'
import ResourceTable from '../components/FilesList/ResourceTable.vue'
import { useResourcesViewDefaults } from '../composables'

import MixinAccessibleBreadcrumb from '../mixins/accessibleBreadcrumb'
import MixinFileActions from '../mixins/fileActions'

import { VisibilityObserver } from 'web-pkg/src/observer'
import { ImageDimension, ImageType } from '../constants'
import debounce from 'lodash-es/debounce'
import { bus } from 'web-pkg/src/instance'

import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import AppBar from '../components/AppBar/AppBar.vue'
import CreateAndUpload from '../components/AppBar/CreateAndUpload.vue'
import NotFoundMessage from '../components/FilesList/NotFoundMessage.vue'
import ListInfo from '../components/FilesList/ListInfo.vue'
import Pagination from '../components/FilesList/Pagination.vue'
import ContextActions from '../components/FilesList/ContextActions.vue'
import { breadcrumbsFromPath, concatBreadcrumbs } from '../helpers/breadcrumbs'
import { defineComponent } from '@vue/composition-api'
import { move } from '../helpers/resource'
import { Resource } from 'web-client'
import { usePublicLinkPassword, useStore } from 'web-pkg/src/composables'
import KeyboardActions from '../components/FilesList/KeyboardActions.vue'
import SideBar from '../components/SideBar/SideBar.vue'
import FilesViewWrapper from '../components/FilesViewWrapper.vue'

const visibilityObserver = new VisibilityObserver()

export default defineComponent({
  components: {
    FilesViewWrapper,
    AppBar,
    CreateAndUpload,
    ResourceTable,
    ListInfo,
    Pagination,
    AppLoadingSpinner,
    NoContentMessage,
    NotFoundMessage,
    ContextActions,
    KeyboardActions,
    SideBar
  },

  mixins: [MixinAccessibleBreadcrumb, MixinFileActions],

  setup() {
    const store = useStore()
    return {
      ...useResourcesViewDefaults<Resource, any, any[]>(),
      publicLinkPassword: usePublicLinkPassword({ store })
    }
  },

  computed: {
    ...mapGetters('Files', [
      'currentFolder',
      'highlightedFile',
      'totalFilesCount',
      'totalFilesSize'
    ]),
    ...mapGetters(['configuration']),

    breadcrumbs() {
      const breadcrumbs = breadcrumbsFromPath(this.$route, this.$route.params.item)
      const rootRoute = breadcrumbs.shift()

      return concatBreadcrumbs(
        { text: this.$gettext('Public link'), to: rootRoute.to },
        ...breadcrumbs
      )
    },

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    folderNotFound() {
      return this.currentFolder === null
    },

    targetRoute() {
      return { name: this.$route.name }
    },

    displayThumbnails() {
      return !this.configuration?.options?.disablePreviews
    }
  },

  watch: {
    $route: {
      handler: function (to, from) {
        const sameRoute = to.name === from?.name
        const sameItem = to.params?.item === from?.params?.item
        if (!sameRoute || !sameItem) {
          this.loadResourcesTask.perform(this, sameRoute)
        }
      },
      immediate: true
    }
  },
  beforeDestroy() {
    visibilityObserver.disconnect()
  },

  created() {
    const loadResourcesEventToken = bus.subscribe('app.files.list.load', (path) => {
      this.loadResourcesTask.perform(this, this.$route.params.item === path, path)
    })

    this.$on('beforeDestroy', () => bus.unsubscribe('app.files.list.load', loadResourcesEventToken))
  },

  methods: {
    ...mapActions('Files', ['loadPreview']),
    ...mapMutations('Files', [
      'SET_CURRENT_FOLDER',
      'LOAD_FILES',
      'CLEAR_CURRENT_FILES_LIST',
      'REMOVE_FILES',
      'REMOVE_FILES_FROM_SEARCHED',
      'REMOVE_FILE_SELECTION'
    ]),

    async fileDropped(fileIdTarget) {
      const selected = [...this.selectedResources]
      const targetInfo = this.paginatedResources.find((e) => e.id === fileIdTarget)
      const isTargetSelected = selected.some((e) => e.id === fileIdTarget)
      if (isTargetSelected) return
      if (targetInfo.type !== 'folder') return
      const movedResources = await move(
        selected,
        targetInfo,
        this.$client,
        this.createModal,
        this.hideModal,
        this.showMessage,
        this.$gettext,
        this.$gettextInterpolate,
        this.$ngettext,
        false,
        this.publicLinkPassword
      )
      for (const resource of movedResources) {
        this.REMOVE_FILES([resource])
        this.REMOVE_FILES_FROM_SEARCHED([resource])
        this.REMOVE_FILE_SELECTION(resource)
      }
    },

    rowMounted(resource, component) {
      if (!this.displayThumbnails) {
        return
      }

      const debounced = debounce(({ unobserve }) => {
        unobserve()
        this.loadPreview({
          resource,
          isPublic: true,
          dimensions: ImageDimension.Thumbnail,
          type: ImageType.Thumbnail
        })
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    }
  }
})
</script>
