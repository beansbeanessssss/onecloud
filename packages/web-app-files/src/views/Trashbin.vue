<template><Trashbin /></template>

<script>
import Trashbin from '../components/Trashbin.vue'

export default {
  components: { Trashbin }
}
</script>
