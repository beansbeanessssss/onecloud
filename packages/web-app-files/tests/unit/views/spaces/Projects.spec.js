import { mount } from '@vue/test-utils'
import { localVue } from '../views.setup'
import { createStore } from 'vuex-extensions'
import mockAxios from 'jest-mock-axios'
import SpaceProjects from '../../../../src/views/spaces/Projects.vue'
import VueRouter from 'vue-router'
import Vuex from 'vuex'

localVue.use(VueRouter)

const selectors = {
  sharesNoContentMessage: '#files-spaces-empty',
  spacesList: '.spaces-list'
}

beforeEach(mockAxios.reset)

describe('Spaces component', () => {
  it('should show a "no content" message', async () => {
    mockAxios.request.mockImplementationOnce(() => {
      return Promise.resolve({
        data: {
          value: []
        }
      })
    })

    const wrapper = getMountedWrapper()
    await wrapper.vm.loadSpacesTask.last

    expect(wrapper.find(selectors.sharesNoContentMessage).exists()).toBeTruthy()
  })

  it('should only list drives of type "project"', async () => {
    mockAxios.request.mockImplementationOnce(() => {
      return Promise.resolve({
        data: {
          value: [
            { driveType: 'project', id: '1' },
            { driveType: 'personal', id: '2' }
          ]
        }
      })
    })

    const wrapper = getMountedWrapper()
    await wrapper.vm.loadSpacesTask.last

    expect(wrapper.vm.spaces.length).toEqual(1)
    expect(wrapper).toMatchSnapshot()
  })

  it('should show the "create new space" modal with sufficient permissions', async () => {
    mockAxios.request.mockImplementationOnce(() => {
      return Promise.resolve({
        data: {
          value: []
        }
      })
    })
    const wrapper = getMountedWrapper()
    await wrapper.vm.loadSpacesTask.last

    const createModalStub = jest.spyOn(wrapper.vm, 'createModal')
    const button = wrapper.find('[data-testid="spaces-list-create-space-btn"]')
    expect(button.exists()).toBeTruthy()
    await button.trigger('click')

    expect(createModalStub).toHaveBeenCalledTimes(1)
  })

  it('should show an error message when trying to create a space with an empty name', async () => {
    mockAxios.request.mockImplementationOnce(() => {
      return Promise.resolve({
        data: {
          value: []
        }
      })
    })
    const wrapper = getMountedWrapper()
    await wrapper.vm.loadSpacesTask.last

    const spyInputErrorMessageStub = jest.spyOn(wrapper.vm, 'setModalInputErrorMessage')
    wrapper.vm.checkSpaceName('')

    expect(spyInputErrorMessageStub).toHaveBeenCalledTimes(1)
  })
})

function getMountedWrapper() {
  return mount(SpaceProjects, {
    localVue,
    router: new VueRouter(),
    store: createStore(Vuex.Store, {
      getters: {
        configuration: () => ({
          server: 'https://example.com/'
        })
      },
      actions: {
        createModal: jest.fn()
      }
    })
  })
}
