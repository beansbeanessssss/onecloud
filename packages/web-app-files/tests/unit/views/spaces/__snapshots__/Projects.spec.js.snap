// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spaces component should only list drives of type "project" 1`] = `
<div class="oc-p-s">
  <h2>Spaces</h2> <span>Access all project related files in one place.</span> <a href="#">Learn more about spaces.</a>
  <h3>Your spaces</h3>
  <hr class="oc-mb-s">
  <div class="spaces-list">
    <div class="
          oc-grid
          oc-grid-match
          oc-grid-column-small
          oc-grid-row-large
          oc-text-center
          oc-child-width-1-3@s
        "><a href="#" class="oc-mb-m"><span class="spaces-list-card oc-border oc-card oc-card-default"><span class="oc-card-media-top oc-border-b"><span class="oc-px-m oc-py-m oc-icon oc-icon-xxl oc-icon-passive"><!----></span></span> <span class="oc-card-body"><span class="oc-card-title"></span></span>
        </span></a>
    </div>
  </div>
</div>
`;
